### ISSUE TITLE
Publisher confirms the right way?

### ISSUE THREAD

**Issue by officer-rosmarino** [squaremo/amqp.node#434]
Hello, I am having problems understanding what is the right way to enable **publisher confirms**. I have the absolute necessity to not lose any message that the publisher sends to the queue.

I am using a worker patter, with one producer, the broker, and many consumers.

As pointed out in the docs [here](http://www.rabbitmq.com/tutorials/tutorial-two-javascript.html):

> **Note on message persistence**
> Marking messages as persistent doesn't fully guarantee that a message won't be lost. Although it tells RabbitMQ to save the message to disk, there is still a short time window when RabbitMQ has accepted a message and hasn't saved it yet. Also, RabbitMQ doesn't do fsync(2) for every message -- it may be just saved to cache and not really written to the disk. The persistence guarantees aren't strong, but it's more than enough for our simple task queue. If you need a stronger guarantee then you can use publisher confirms.

I gave a look at the amqp library docs and I have found that I can create channels like this `conn.createConfirmChannel();` and I am wondering if this is all it takes to enable publisher confirms and be sure that messages sent by the publisher to the broker will not be lost.

**Comment by squaremo** [comment]
That is all it takes to enable publisher confirms.

They are not in themselves a guarantee that nothing will be lost. They will tell you if a message has _definitely_ been enqueued (including being fsync'd to disk, if the queue is durable and the message is persistent), or _definitely not_ been enqueued.

You still have to decide what to do in the case that a message has _definitely not_ been enqueued, and what to do in the case that you don't hear back at all (or don't hear back within a timeout). If you absolutely cannot lose a message, then you probably want to resend, and deduplicate at the far end (if it matters that you repeat a job).

**Comment by cressie176** [comment]
Pretty sure you don't need waitForConfirms - see [here](https://github.com/squaremo/amqp.node/issues/375)

**Comment by cressie176** [comment]
WRT to your question, you say you absolutely can't afford to lose any messages. You've used a confirm channel, made the queue durable and the message persistent. My understanding from the [amqplib](http://www.squaremobius.net/amqp.node/channel_api.html#flowcontrol) and [RabbitMQ](https://www.rabbitmq.com/confirms.html#publisher-confirms) docs is that once the ch.sendToQueue returns true, you are guaranteed that the message was delivered to the broker, and written to disk for any queues it will be routed to. If you've messed up your routing / queue name, the message will be dropped but you'll still get a confirm.

The gotcha is that there's no way to tell how long this will take, so your application could stop/crash/die before the message has been published, or after it's been published, but before the confirm was returned. If you really can't afford to lose messages, you'll need the ability to republish in this event, and will have to design your system to tolerate duplicates.

There are other ways you can lose messages. If you only have a single broker, you could lose the disk. If you're running a cluster of brokers, and the queue is replicated you can get a network partition and lose messages as described towards the end of this [article](https://aphyr.com/posts/315-jepsen-rabbitmq).

In summary, if you really, really can't afford to lose messages with RabbitMQ you need to build the facility to work out which messages you may have lost, republish them and tolerate duplicates. A few other brokers offer stronger guarantees, but come with different sets of problems, that are often glossed over or overlooked in the documentation and blog posts. 

Most of the applications I write have natural support for republishing - there's a human at one end who can click the button again. I've yet to find a combination of broker and client I like as much as amqplib and rabbitmq, so I'm not sure what I'll do when I genuinely have this problem.

**Comment by honestserpent** [comment]
@cressie176 thank you for your answer. What I needed is a guarantee that once I send a message to the queue and it returns, the message has definetly been enqueued and f'syncd to the disk.

**Comment by MartinWickman** [comment]
You can use transactions with RabbitMQ. According to the rabbit docs, you should be able to send "txSelect" to rabbit which will then put the channel into transactions mode. This is slow as hell to use, but it should guarantee that the message is written to disk.

 I couldn't find anything about it in the amqplib api docs though?

**Comment by honestserpent** [comment]
@MartinWickman Yeah I can't find complete examples. From what I have understood so far, in order to be sure your message _has definetly been enqueued and fsync'd_,  it is enough to use a ConfirmChannel.

I still haven't quite understood if I need to use _waitForConfirms_ and if yes, what for.

**Comment by honestserpent** [comment]
Ok, so basically, say I have only one message. If I get it right, I could either do this:

```
let res = ch.sendToQueue('queueName', Buffer.from({a:'data'}), {
   persistent: true
}, function (err, ok) {
  if(err) ...
});
```
or
```
let res = ch.sendToQueue('queueName', Buffer.from({a:'data'}), {persistent: true});
ch.waitForConfirms(function(err){
  if(err) ...
});
```
and there is no difference.
The difference is if I have multiple messages. Am I right?

**Comment by salamander-mh** [comment]
Note `sendToQueue` Does not return a promise

**Comment by alexchen1875** [comment]
Is there a place to find a complete example，how to use confirmsChannel ?

**Comment by cressie176** [comment]
I created [this](https://github.com/cressie176/amqplib-express-example/blob/master/README.md). It doesn't use send to queue, as I always prefer the flexibility of publishing to an exchange. Going to close this issue as it's been open for a while, and I don't think there's much more to add.
